api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirPatientDaoImpl.java

```diff
@@ -104,4 +114,53 @@ protected String paramToProp(String param) {
 		
 		return super.paramToProp(param);
 	}
+	
+	private void handleNames(List<PropParam<?>> params, Criteria criteria) {
+		StringAndListParam name = null;
+		StringAndListParam given = null;
+		StringAndListParam family = null;
+		for (PropParam<?> param : params) {
+			switch (param.getPropertyName()) {
+				case FhirConstants.NAME_PROPERTY:
+					name = (StringAndListParam) param.getParam();
+					break;
+				case FhirConstants.GIVEN_PROPERTY:
+					given = (StringAndListParam) param.getParam();
+					break;
+				case FhirConstants.FAMILY_PROPERTY:
+					family = (StringAndListParam) param.getParam();
+					break;
+			}
+		}
+		
+		handleNames(criteria, name, given, family);
+	}
+	
+	private void handleAddresses(Criteria criteria, Map.Entry<String, List<PropParam<?>>> entry) {
+		StringAndListParam city = null;
+		StringAndListParam country = null;
+		StringAndListParam postalCode = null;
+		StringAndListParam state = null;
+		for (PropParam<?> param : entry.getValue()) {
+			switch (param.getPropertyName()) {
+				case FhirConstants.CITY_PROPERTY:
+					city = ((StringAndListParam) param.getParam());
+					break;
+				case FhirConstants.COUNTRY_PROPERTY:
+					country = ((StringAndListParam) param.getParam());
+					break;
+				case FhirConstants.POSTAL_CODE_PROPERTY:
+					postalCode = ((StringAndListParam) param.getParam());
+					break;
+				case FhirConstants.STATE_PROPERTY:
+					state = ((StringAndListParam) param.getParam());
+					break;
```

Comment:
Really minor point, but could we order these city, state, postal code, property (that's roughly ascending order of size (except that postal code probably belong at the front of the list) just to keep things consistent (my main hang-up is making sure "state" is before "country" since it isn't always obvious to people that "state" means state in the sense of "US state" or "Australian state" rather than "country").